File: src/main/java/com/claudijusapchy/ratprotection/mixin/UserMixin.java
package com.claudijusapchy.ratprotection.mixin;

import com.claudijusapchy.ratprotection.ModLogger;
import net.minecraft.client.User;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
import java.util.List;

@Mixin(User.class)
public class UserMixin {

    private static final List<String> TRUSTED = List.of(
            "net.minecraft", "net.fabricmc", "com.mojang",
            "com.claudijusapchy.ratprotection", "java.", "sun.",
            "jdk.", "kotlin.", "org.spongepowered", "org.prismlauncher",
            "net.caffeinemc", "me.owdding"
    );

    @Inject(method = "getAccessToken", at = @At("HEAD"))
    private void onGetAccessToken(CallbackInfoReturnable<String> cir) {
        checkCaller("getAccessToken");
    }

    @Inject(method = "getProfileId", at = @At("HEAD"))
    private void onGetProfileId(CallbackInfoReturnable<java.util.UUID> cir) {
        checkCaller("getProfileId");
    }

    private void checkCaller(String method) {
        for (StackTraceElement frame : Thread.currentThread().getStackTrace()) {
            String cls = frame.getClassName();
            if (TRUSTED.stream().anyMatch(cls::startsWith)) continue;
            if (cls.contains("UserMixin") || cls.contains("ratprotection")) continue;
            ModLogger.INSTANCE.block("[RatProtection] ALERT: Suspicious access to " + method + " by " + cls);
            return;
        }
    }
}

File: src/main/mixin/UserMixin.kt
package com.claudijusapchy.ratprotection.mixin

import com.claudijusapchy.ratprotection.ModLogger
import net.minecraft.client.User
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(User::class)
class UserMixin {

    // Trusted packages that are allowed to access the token
    private val trustedPackages = listOf(
        "net.minecraft",
        "net.fabricmc",
        "com.mojang",
        "com.claudijusapchy.ratprotection",
        "java.",
        "sun.",
        "jdk.",
        "kotlin.",
        "org.spongepowered"
    )

    @Inject(method = ["getAccessToken"], at = [At("HEAD")])
    fun onGetAccessToken(cir: CallbackInfoReturnable<String>) {
        checkCaller("getAccessToken")
    }

    @Inject(method = ["getProfileId"], at = [At("HEAD")])
    fun onGetProfileId(cir: CallbackInfoReturnable<Any>) {
        checkCaller("getProfileId")
    }

    private fun checkCaller(method: String) {
        val stack = Thread.currentThread().stackTrace

        // Walk up the call stack looking for suspicious callers
        for (frame in stack) {
            val className = frame.className

            // Skip trusted packages
            if (trustedPackages.any { className.startsWith(it) }) continue

            // Skip the mixin itself
            if (className.contains("UserMixin")) continue
            if (className.contains("ratprotection")) continue

            // Anything else calling this is suspicious
            ModLogger.block("[RatProtection] ALERT: Suspicious mod accessing $method!")
            ModLogger.block("[RatProtection] Caller: $className.${frame.methodName}(${frame.fileName}:${frame.lineNumber})")
            return
        }
    }
}